url063b9533f6_karlney_tetris_ai_tgz-pJ8-karlney_tetris_cmd_TetrisGameRunnerJ8/src/karlney/tetris/core/TetrisGame.java

```diff
@@ -0,0 +1,106 @@
+package karlney.tetris.core;
+
+import java.util.List;
+
+/**
+ * TODO javadoc
+ * <p/>
+ * Date: 2012-11-01
+ * Time: 20:36
+ *
+ * @author karl.neyvaldt
+ */
+public class TetrisGame {
+
+    // http://colinfahey.com/tetris/tetris.html
+    private static final int MAX_LEVEL = 10; //level 11 is only for AI training
+    private static final int MIN_LEVEL = 1;  //level 0 is only for debugging
+
+    private static final int[] LEVEL_DELAYS = {Integer.MAX_VALUE, 500,450,400,350,300,250,200,150,100,50,0};
+
+    //True when the game is started (False only if the game is created but not yet started)
+    private boolean gameStarted;
+
+    //True when the game is paused
+    private boolean paused;
+
+    //Current level
+    private int level;
+
+    //The players
+    private List<Player> players;
+
+    public TetrisGame(int startLevel, List<Player> players) {
+        this.level = startLevel;
+        this.players = players;
+        this.gameStarted = false;
+    }
+
+    public void start(){
+        if (!gameStarted){
+            gameStarted = true;
+            for (Player pl:players) {
+                pl.start(getDelay());
+            }
+        }
+    }
+
+    public void stop(){
+        for (Player pl:players) {
+            pl.stop();
+        }
+    }
+
+    public void increaseLevel(){
+        level++;
+        if	(level>= MAX_LEVEL)	level= MAX_LEVEL;
+        for (Player pl:players) {
+            pl.setDelay(getDelay());
+            pl.setLevel(level);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 55:60, 64:69.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.